training_engine_sae.py

```diff
@@ -0,0 +1,152 @@
+    conv8 = Conv2D(ff * 4, 3, activation='relu', padding='same', kernel_initializer='he_normal')(conv8)
+    up9 = UpSampling2D(size=(2, 2))(conv8)
+    up9 = Conv2D(ff * 2, 2, activation='relu', padding='same', kernel_initializer='he_normal')(up9)
+    merge9 = Concatenate(axis = 3)([conv1,up9])
+
+    conv9 = Conv2D(ff * 2, 3, activation='relu', padding='same', kernel_initializer='he_normal')(merge9)
+    conv9 = Conv2D(ff * 2, 3, activation='relu', padding='same', kernel_initializer='he_normal')(conv9)
+    conv9 = Conv2D(2, 3, activation='relu', padding='same', kernel_initializer='he_normal')(conv9)
+    conv10 = Conv2D(1, 1, activation='sigmoid')(conv9)
+
+    model = Model(inputs=inputs, outputs=conv10)
+
+    model.compile(optimizer=Adam(lr=1e-4), loss='binary_crossentropy', metrics=['accuracy'])
+
+    if pretrained_weights is not None:
+        model.load_weights(pretrained_weights)
+
+    return model
+
+
+def getTrain(input_image, gt, patch_height, patch_width):
+    X_train = []
+    Y_train = {}
+
+    hstride = patch_height // 2
+    wstride = patch_width // 2
+
+    # TODO Take into account margins
+    for h in range(0, input_image.shape[0] - patch_height, hstride):
+        for w in range(0, input_image.shape[1] - patch_width, wstride):
+            x_sample = input_image[h:h + hspan, w:w + wspan]
+
+            # Pre-process TODO: Check all the pipelines to do the same
+            x_sample = (255. - x_sample) / 255.
+
+            X_train.append(x_sample)
+
+            for label in gt:
+                Y_train[label].append(gt[label][h:h + hspan, w:w + wspan])
+
+    return X_train, Y_train
+
+
+def train_msae(input_image, gt, height, width, output_path, epochs):
+    # -------------------------------------------------------------------------------------------------------------------
+
+    # Crop the image and create ground_truth
+    [X_train, Y_train] = getTrain(input_image, gt,
+                                  height, width)
+
+    print 'Training created with ' + str(len(X_train)) + ' samples.'
+    for label in Y_train:
+
+        # Training configuration
+        print 'Training a new model for ' + str(label)
+        model = get_sae(
+            height=height,
+            width=width
+        )
+
+        model.summary()
+
+        callbacks_list = [
+                ModelCheckpoint(output_path[label], save_best_only=True, monitor='val_acc', verbose=1, mode='max'),
+                EarlyStopping(monitor='val_acc', patience=3, verbose=0, mode='max')
+                ]
+
+        # Training stage
+        model.fit(X_train, Y_train[label],
+                  verbose=2,
+                  batch_size=BATCH_SIZE,
+                  validation_split=VALIDATION_SPLIT,
+                  callbacks=callbacks_list,
+                  epochs=epochs)
+
+    return 0
+
+# Debuggin code
+if __name__ == "__main__":
+    input_path = '/Users/jcalvo/Desktop/McGill/dama-corpora/Corpora/Layers/Salzinnes/CF-009/'
```

Comment:
Instead of hardcoding a path, you may want to leave a set of testing images in the repo. Or write a unit test :)

Reply:
I think this was already removed in the last commit (it was for my own testing).